### PR TITLE
mount: restart mount process gracefully

### PR DESCRIPTION
**Pull request by davies**

### CODE REVIEW

**Comment by zhijian-pro** [comment]
Badger does not support smooth upgrades because only one database connection can be active at a time.

File: pkg/meta/base.go [hunk 1]
```diff
@@ -475,23 +476,28 @@ func (m *baseMeta) newSessionInfo() []byte {
...
+			if err != nil {
+				return 0, fmt.Errorf("get session ID: %s", err)
+			}
+			m.sid = uint64(v)
+			action = "Create"
 		}
-		m.sid = uint64(v)
-		if err = m.en.doNewSession(m.newSessionInfo()); err != nil {
-			return fmt.Errorf("create session: %s", err)
+		if err := m.en.doNewSession(m.newSessionInfo()); err != nil {
```
Comment (davies) [inline pkg/meta/base.go:497]: should use existing session

File: pkg/fuse/fuse.go [hunk 2]
```diff
@@ -501,6 +502,53 @@ func Serve(v *vfs.VFS, options string, xattrs, ioctl bool) error {
...
+	for _, n := range strings.Split(options, ",") {
+		// TODO allow_root
+		if n == "allow_other" {
+			opt.AllowOther = true
+		} else if strings.HasPrefix(n, "fsname=") {
+			opt.FsName = n[len("fsname="):]
+		} else if n == "debug" {
+			opt.Debug = true
+			log.SetFlags(log.Ldate | log.Ltime | log.Lmicroseconds)
+		} else if strings.HasPrefix(n, "mfs") {
```
Comment (davies) [inline pkg/fuse/fuse.go:533]: remove it

File: cmd/mount_unix.go [hunk 3]
```diff
@@ -146,6 +240,14 @@ func fuseFlags() []cli.Flag {
 			Name:  "enable-ioctl",
 			Usage: "enable ioctl (support GETFLAGS/SETFLAGS only)",
 		},
+		&cli.BoolFlag{
+			Name:  "no-bsd-lock",
```
Comment (davies) [inline cmd/mount_unix.go:244]: remove it

File: pkg/vfs/vfs.go [hunk 4]
```diff
@@ -48,6 +50,63 @@ type Port struct {
...
+		o.Options = append(o.Options, opt)
+	}
+
+	sort.Strings(o.Options)
+
+	// ignore these options because they won't be send to kernel
+	o.IgnoreSecurityLabels,
+		o.RememberInodes,
+		o.SingleThreaded,
+		o.DisableXAttrs,
```
Comment (davies) [inline pkg/vfs/vfs.go:96]: Once xattr is disabled, ENOSYS will be returned to kernel, then no more request will be sent to FUSE

File: pkg/vfs/vfs.go [hunk 5]
```diff
@@ -1031,6 +1104,22 @@ func NewVFS(conf *Config, m meta.Meta, store chunk.ChunkStore, registerer promet
 		meta.TrashName = ".jfs" + meta.TrashName
 	}
 
+	statePath := os.Getenv("_FUSE_STATE_PATH")
+	if statePath == "" {
+		statePath = fmt.Sprintf("/tmp/state%d.json", os.Getppid())
+	} else {
+		// there is a bug in the supervsior, _FUSE_STATE_PATH could be stale
```
Comment (davies) [inline pkg/vfs/vfs.go:1111]: remove these

File: pkg/vfs/handle.go [hunk 6]
```diff
@@ -235,3 +241,123 @@ func (v *VFS) releaseFileHandle(ino Ino, fh uint64) {
...
+			}
+			s := saveHandle{
+				Inode:      uint64(h.inode),
+				Length:     length,
+				Flags:      h.flags,
+				UseLocks:   h.locks,
+				FlockOwner: h.flockOwner,
+				Off:        h.off,
+				Data:       hex.EncodeToString(h.data),
+				Pending:    hex.EncodeToString(h.pending),
```
Comment (davies) [inline pkg/vfs/handle.go:288]: pending is only used by `controlInode`, remove it

**Pull request merged**